india_compliance/gst_india/utils/e_invoice.py

```diff
@@ -112,6 +116,81 @@ def log_error():
             frappe.db.commit()  # nosemgrep
 
 
+@frappe.whitelist()
+def handle_duplicate_irn_error(data, current_gstin, current_invoice_amount, **kwargs):
+    """
+    Handle Duplicate IRN errors by fetching the IRN details and comparing with the current invoice.
+
+    Steps:
+    1. Fetch IRN details using the IRN number using e-Invoice API.
+    2. If the IRN details cannot be fetched, fetch the IRN details from the GST Portal.
+    3. Compare the buyer GSTIN and invoice amount with the current invoice and throw an error if they don't match.
+    """
+
+    if isinstance(data, str):
+        data = frappe._dict(json.loads(data))
+
+    if isinstance(current_invoice_amount, str):
+        current_invoice_amount = flt(current_invoice_amount)
+
+    docname = kwargs.get("docname")
+
+    doc = kwargs.get("doc") or load_doc("Sales Invoice", docname, "submit")
+    settings = kwargs.get("settings") or frappe.get_cached_doc("GST Settings")
+    api = kwargs.get("api") or EInvoiceAPI(doc)
+
+    response = api.get_e_invoice_by_irn(data.Irn)
+
+    # Handle error 2283:
+    # IRN details cannot be provided as it is generated more than 2 days ago
+    if (
+        response.error_code == "2283"
+        and settings.fetch_e_invoice_details_from_gst_portal
+    ):
+        response = TaxpayerEInvoiceAPI(doc).get_irn_details(data.Irn)
+
+        if response.error_type == "otp_requested":
+            response.update(
+                {
+                    "data": data,
+                    "docname": docname,
+                    "current_gstin": current_gstin,
+                    "current_invoice_amount": current_invoice_amount,
+                }
+            )
+
+            return response
+
+        response = frappe._dict(response.data or response.error)
+
+    if signed_data := response.SignedInvoice:
+        invoice_data = json.loads(
+            jwt.decode(signed_data, options={"verify_signature": False})["data"]
+        )
+
+        previous_gstin = invoice_data.get("BuyerDtls").get("Gstin")
+        previous_invoice_amount = invoice_data.get("ValDtls").get("TotInvVal")
+
+        if not (
+            previous_gstin == current_gstin
+            and previous_invoice_amount == current_invoice_amount
+        ):
+            frappe.throw(
+                _(
+                    "e-Invoice is already available against Invoice {0} with a Grand Total of Rs.{1}"
+                    " Duplicate IRN requests are not considered by e-Invoice Portal."
```

Comment:
Hence, IRN number {} is not updated against current Invoice.

Error:
1) Total amount is different
2) Customer GSTIN is different

Solution:
1) Use different Invoice Number if you are generating new invoice or amendment is done after generation of e-Invoice.
2) Try cancelling e-Invoice from e-Invoice portal if old invoice is invalid and is generate in last 24 hrs.